tests/README.md

```diff
@@ -0,0 +1,386 @@
+
+    $ make
+
+
+Run one test at a time:
+
+    $ make run_seq
+
+
+Run a single test:
+
+    $ TESTS=0034 make
+
+
+Run test suite with valgrind (see instructions below):
+
+    $ ./run-test.sh ./merged valgrind   # memory checking
+
+or with helgrind (the valgrind thread checker):
+
+    $ ./run-test.sh ./merged helgrind   # thread checking
+
+
+To run the tests in gdb:
+
+**NOTE**: gdb support is flaky on OSX due to signing issues.
+
+    $ ./run-test.sh ./merged gdb
+    (gdb) run
+
+    # wait for test to crash, or interrupt with Ctrl-C
+
+    # backtrace of current thread
+    (gdb) bt
+    # move up or down a stack frame
+    (gdb) up
+    (gdb) down
+    # select specific stack frame
+    (gdb) frame 3
+    # show code at location
+    (gdb) list
+
+    # print variable content
+    (gdb) p rk.rk_conf.group_id
+    (gdb) p *rkb
+
+    # continue execution (if interrupted)
+    (gdb) cont
+
+    # single-step one instruction
+    (gdb) step
+
+    # restart
+    (gdb) run
+
+    # see all threads
+    (gdb) info threads
+
+    # see backtraces of all threads
+    (gdb) thread apply all bt
+
+    # exit gdb
+    (gdb) exit
+
+
+If a test crashes and produces a core file (make sure your shell has
+`ulimit -c unlimited` set!), do:
+
+    # On linux
+    $ LD_LIBRARY_PATH=../src:../src-cpp gdb ./merged <core-file>
+    (gdb) bt
+
+    # On OSX
+    $ DYLD_LIBRARY_PATH=../src:../src-cpp gdb ./merged /cores/core.<pid>
+    (gdb) bt
+
+
+To run all tests repeatedly until one fails, this is a good way of finding
+intermittent failures, race conditions, etc:
+
+    $ ./until-fail.sh bare  # bare is to run the test without valgrind,
+                            # may also be one or more of the modes supported
+                            # by run-test.sh:
+                            #  bare valgrind helgrind gdb, etc..
+
+To run a single test repeatedly with valgrind until failure:
+
+    $ TESTS=0103 ./until-fail.sh valgrind
+
+
+
+### Finding memory leaks, memory corruption, etc.
+
+There are two ways to verifying there are no memory leaks, out of bound
+memory accesses, use after free, etc. ASAN or valgrind.
+
+#### ASAN - AddressSanitizer
+
+The first option is using AddressSanitizer, this is build-time instrumentation
+provided by clang and gcc to insert memory checks in the build library.
+
+To enable AddressSanitizer (ASAN), run `./dev-conf.sh asan` from the
+librdkafka root directory.
+This script will rebuild librdkafka and the test suite with ASAN enabled.
+
+Then run tests as usual. Memory access issues will be reported on stderr
+in real time as they happen (and the test will fail eventually), while
+memory leaks will be reported on stderr when the test run exits successfully,
+i.e., no tests failed.
+
+Test failures will typically cause the current test to exit hard without
+cleaning up, in which case there will be a large number of reported memory
+leaks, these shall be ignored. The memory leak report is only relevant
```

Comment:
in my experience ASAN simply doesn't print anything on test failure. Valgrind will report a bunch of extra junk but ASAN remained silent.

Reply:
It depends.